_keystone/saml-federation.md

```diff
@@ -0,0 +1,541 @@
+   1. The IDP supports the SAML ECP protocol
+   2. The IDP supports basic authentication on a GET
+
+The CLI offers two types of plugins:
+  1. Unscoped (To get an unscoped token)
+  2. Scoped
+
+1. Unscoped Example:
+
+Let's create a file that we can source in our environment to use SAML:
+
+```
+# This file is called companyxyz_saml_rc
+export OS_IDENTITY_API_VERSION=3
+export OS_PASSWORD=CompanyXYZPassword #replace this
+export OS_AUTH_URL=https://<BB CLOUD FQDN>:5000/v3 #replace this
+export OS_USERNAME=CompanyXYZUser #replace this
+export OS_IDENTITY_PROVIDER=company-xyz-idp #replace this
+export OS_PROTOCOL=saml2
+export OS_AUTH_TYPE=v3unscopedsaml
+export OS_IDENTITY_PROVIDER_URL=https://CompanyXYZ.example/idp/profile/SAML2/SOAP/ECP #replace this
+```
+
+After creating the file and editing the options we can run the following command to get an unscoped token and record the ID:
+
+```
+$ source companyxyz_saml_rc
+$ openstack token issue
+```
+
+2. Scoped Example:
+
+Now that we a have an unscoped token, we can scope to a project and list the servers on that project
+
+```
+openstack server list --os-auth-type v3scopedsaml --os-token <os-token-id> --os-project-id <project-id>
+```
+
+## <a name="saml"></a>Using the Python API Libraries
+
+**Note: The Python API Libraries can only work with IDPs that satisfy these conditions:**
+   1. The IDP supports the SAML ECP protocol
+   2. The IDP supports basic authentication on a GET
+
+The following code uses the SAML auth plugin to authenticate with the IDP,
+send the SAML assertion to Keystone and retreive an unscoped token.
+THe unscoped token is then used to get a project scoped token. Then the script
+will list servers on that project.
+
+This file is saml_list_servers.py:
+
+{% highlight bash %}
+#!/usr/bin/python
+
+from keystoneauth1 import session
+from keystoneauth1.identity.v3 import Token
+from keystoneauth1.extras._saml2.v3.saml2 import Password as SAMLPassword
+from keystoneclient.v3.client import Client
+
+from novaclient import client
+import os
+
+
+# we get these variables from sourcing a file
+OS_AUTH_URL = os.environ.get('OS_AUTH_URL')
+OS_USERNAME = os.environ.get('OS_USERNAME')
+OS_PASSWORD = os.environ.get('OS_PASSWORD')
+OS_PROJECT_NAME = os.environ.get('OS_PROJECT_NAME')
+OS_IDENTITY_PROVIDER = os.environ.get('OS_IDENTITY_PROVIDER')
+OS_IDENTITY_PROVIDER_URL = os.environ.get('OS_IDENTITY_PROVIDER_URL')
+OS_PROTOCOL = os.environ.get('OS_PROTOCOL', 'saml2')
+
+
+def get_project_and_federated_session():
+    # authenticate to the idp
+    saml_auth = SAMLPassword(auth_url=OS_AUTH_URL,
+                             identity_provider=OS_IDENTITY_PROVIDER,
+                             identity_provider_url=OS_IDENTITY_PROVIDER_URL,
+                             username=OS_USERNAME,
+                             password=OS_PASSWORD,
+                             protocol=OS_PROTOCOL)
```

Comment:
I have updated the formatting for other parts of the page. Also, the page appearing on github greatly differs on what it looks like in the end. There is an explanation on how to run a local server here: https://github.com/IBM-Blue-Box-Help/help-documentation/pull/164